libraries/extensions/ros2-bridge/python/src/typed/serialize/mod.rs

```diff
@@ -0,0 +1,172 @@
+use std::{borrow::Cow, collections::HashMap, fmt::Display};
+
+use arrow::array::{Array, ArrayRef, AsArray};
+use dora_ros2_bridge_msg_gen::types::{
+    primitives::{GenericString, NestableType},
+    MemberType,
+};
+use serde::ser::SerializeTupleStruct;
+
+use super::{TypeInfo, DUMMY_STRUCT_NAME};
+
+mod array;
+mod defaults;
+mod primitive;
+mod sequence;
+
+#[derive(Debug, Clone)]
+pub struct TypedValue<'a> {
+    pub value: &'a ArrayRef,
+    pub type_info: &'a TypeInfo<'a>,
+}
+
+impl serde::Serialize for TypedValue<'_> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let empty = HashMap::new();
+        let package_messages = self
+            .type_info
+            .messages
+            .get(self.type_info.package_name.as_ref())
+            .unwrap_or(&empty);
+        let message = package_messages
+            .get(self.type_info.message_name.as_ref())
+            .ok_or_else(|| {
+                error(format!(
+                    "could not find message type {}::{}",
+                    self.type_info.package_name, self.type_info.message_name
+                ))
+            })?;
+
+        let input = self
+            .value
+            .as_struct_opt()
+            .ok_or_else(|| error("expected struct array"))?;
```

Comment:
Makes sense, thanks!

Do you think that printing `self.value.data_type()` instead of `self.value` would be enough? Printing the whole value could lead to a lot of output for large messages, e.g. when sending image data.

Reply:
Well, the thing is sometimes, we have many of the same datatype like `UInt8` and it might be hard to know which one is the issue. But I agree that `self.value` can be very big.

Reply:
I'm not sure if I understand. This error occurs when we try to cast the input as a struct array, but the input array is of some other type. So the array contents don't seem relevant?

Reply:
Well, it is sometimes hard to know which member of the message ( more so if the member is nested ) should be a struct and was cast as something else. 

I guess it is additional information. But, I agree that it is not required.

We can use `self.value.data_type()`

Reply:
Thanks for clarifying!

I just pushed [1addfa3](https://github.com/dora-rs/dora/pull/415/commits/1addfa385a4815a32e817423f99a933f403a4fc1) to switch to `self.value.data_type()`. To make debugging easier, I added more error context when trying to serialize a struct field. Now the messages look like the following:

> RuntimeError: publish failed
>
> Caused by:
>    Serialization error: Serde says:failed to serialize field Twist.angular: Serde says:given struct has unknown field foo

The "Serde says:" part is coming directly from `rustdds`, so we cannot remove it unfortunately.